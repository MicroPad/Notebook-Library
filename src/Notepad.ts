import { format, parse } from 'date-fns';
import { Builder } from 'xml2js';
import FlatNotepad, { FlatSection } from './FlatNotepad';
import Note, { MarkdownNote } from './Note';
import { NotepadShell } from "./interfaces";
import { encrypt, EncryptionMethod } from './crypto';
import { LAST_MODIFIED_FORMAT } from './date-formats';
import Asset from './Asset';
import Section from './Section';

export type NotepadOptions = {
	lastModified?: Date;
	sections?: Section[];
	notepadAssets?: string[];
	assets?: Asset[];
	crypto?: EncryptionMethod;
};

/**
 * This class is identical in structure to the old Notepad class from the original parser.
 * This represents the notepad as a tree. If you're looking for a flatter structure you can use {@link FlatNotepad}.
 *
 * Something to remember is that all operations on this class like addSection, will return a <strong>new</strong>
 * object of this class, and not modify the existing one.
 */
export default class Notepad implements NotepadShell {
	public readonly lastModified: string;
	public readonly sections: Section[];
	public readonly notepadAssets: string[];
	public readonly assets: Asset[];
	public readonly crypto?: EncryptionMethod;

	constructor(
		public readonly title: string,
		opts: NotepadOptions = {}
	) {
		this.lastModified = format(opts.lastModified || new Date(), LAST_MODIFIED_FORMAT);
		this.sections = opts.sections || [];
		this.notepadAssets = opts.notepadAssets || [];
		this.assets = opts.assets || [];
		if (opts.crypto) this.crypto = opts.crypto;
	}

	public addSection(section: Section): Notepad {
		const notepad = this.clone({
			sections: [
				...this.sections,
				section
			]
		});
		section.parent = notepad;

		return notepad;
	}

	public addAsset(asset: Asset): Notepad {
		return this.clone({
			assets: [
				...this.assets,
				asset
			],
			notepadAssets: [
				...this.notepadAssets,
				asset.uuid
			]
		});
	}

	/**
	 * This updates the lastModified date on the notepad. This date is used for syncing so it's important
	 * to call this method whenever a change is made that will need to be synced.
	 * @param {Date} lastModified
	 * @returns {Notepad}
	 */
	public modified(lastModified: Date = new Date()): Notepad {
		return this.clone({
			lastModified
		});
	}

	/**
	 * @param {string} query Can either be a title-search or a hashtag-search
	 * @returns {Note[]}
	 */
	public search(query: string): Note[] {
		return this.sections
			.map(s => s.search(query))
			.reduce((acc, val) => acc.concat(val), []);
	}

	public async toJson(passkey?: string): Promise<string> {
		let notepad: Notepad = {
			...this,
			assets: undefined,

			// If we're given a passkey but encryption hasn't been setup on the notepad, set it up
			crypto: (!this.crypto && !!passkey) ? 'AES-256-GZ' : this.crypto
		};

		let notepadToStringify: NotepadShell = notepad;
		if (!!notepad.crypto && !!passkey) notepadToStringify = await encrypt(notepad, passkey);

		if (!!notepad.crypto && !passkey) {
			notepadToStringify = { ...notepadToStringify, crypto: undefined };
		}

		return JSON.stringify(notepadToStringify, (key, value) => {
			return (key === 'parent') ? undefined : value;
		});
	}

	public async toXml(): Promise<string> {
		const builder = new Builder({
			cdata: true,
			renderOpts: {
				'pretty': false
			},
			xmldec: {
				version: '1.0',
				encoding: 'UTF-8',
				standalone: false
			}
		});

		// Generate the XML
		const obj = await this.toXmlObject();
		return builder.buildObject(obj).replace(/&#xD;/g, '');
	}

	public flatten(): FlatNotepad {
		let notepad = new FlatNotepad(this.title, {
			lastModified: parse(this.lastModified, LAST_MODIFIED_FORMAT, new Date()),
			notepadAssets: this.notepadAssets,
			crypto: this.crypto
		});

		const flattenSection = (section: Section) => {
			let flat: FlatSection = { title: section.title, internalRef: section.internalRef };
			if (section.parent) flat.parentRef = (section.parent as Section).internalRef;

			// Add this flat section
			notepad = notepad.addSection(flat);
			section.notes.forEach(n => notepad = notepad.addNote(n));

			// Add all of its children recursively
			section.sections.forEach(s => flattenSection(s));
		};
		this.sections.forEach(s => flattenSection(s));

		return notepad;
	}

	public async toMarkdown(assets: Asset[]): Promise<MarkdownNote[]> {
		return (
			await Promise.all(
				this.sections.map(s => s.toMarkdown(assets))
			)
		).reduce((acc, val) => acc.concat(val), []);
	}

	public clone(opts: Partial<NotepadOptions> = {}, title: string = this.title): Notepad {
		return new Notepad(title, {
			lastModified: parse(this.lastModified, LAST_MODIFIED_FORMAT, new Date()),
			sections: [...this.sections],
			notepadAssets: [...this.notepadAssets],
			assets: [...this.assets],
			crypto: this.crypto,
			...opts
		});
	}

	private async toXmlObject(): Promise<object> {
		return {
			notepad: {
				$: {
					'xsi:noNamespaceSchemaLocation': 'https://getmicropad.com/schema.xsd',
					title: this.title,
					lastModified: this.lastModified,
					'xmlns:xsi': 'http://www.w3.org/2001/XMLSchema-instance'
				},
				assets: [
					{
						asset: await Promise.all(this.assets.map(a => a.toXmlObject()))
					}
				],
				section: this.sections.map(s => s.toXmlObject().section)
			}
		};
	}
}
