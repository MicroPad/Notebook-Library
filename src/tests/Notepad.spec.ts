import { Asset, Notepad, Section } from '../';
import { NotepadOptions } from '../Notepad';
import { TestUtils } from './TestUtils';

describe('Notepad', () => {
	let options = getOptions();

	beforeEach(() => {
		options = getOptions();
	});

	describe('constructor', () => {
		it('should construct with just a title', () => {
			// Arrange
			const title = 'test';

			// Act
			const n = new Notepad(title);

			// Assert
			expect(n).toBeInstanceOf(Notepad);
			expect(n.title).toEqual(title);
		});

		Object.entries(options).forEach(option =>
			it(`should construct with ${option[0]}`, () => {
				// Arrange
				const title = 'test';

				// Act
				const n = new Notepad(title, {
					[option[0]]: option[1]
				});

				// Assert
				expect(n).toBeInstanceOf(Notepad);
				expect(n.title).toEqual(title);
				expect(n[option[0]]).toMatchSnapshot();
			})
		);
	});

	describe('addSection', () => {
		let notepad: Notepad;
		let section: Section;

		beforeEach(() => {
			notepad = new Notepad('test');
			section = new Section('test');
		});

		it('should add a new section', () => {
			//Arrange
			// Act
			const res = notepad.addSection(section);

			// Assert
			expect(res.sections[0]).toEqual(section);
		});

		it('should create a new object', () => {
			//Arrange
			// Act
			const res = notepad.addSection(section);

			// Assert
			expect(res).not.toBe(notepad);
		});
	});

	describe('addAsset', () => {
		let notepad: Notepad;
		let asset: Asset;

		beforeEach(() => {
			notepad = new Notepad('test');
			asset = TestUtils.makeAsset();
		});

		it('should add a new asset', () => {
			//Arrange
			// Act
			const res = notepad.addAsset(asset);

			// Assert
			expect(res.assets[0]).toEqual(asset);
		});

		it('should add a new value to notepadAssets', () => {
			//Arrange
			// Act
			const res = notepad.addAsset(asset);

			// Assert
			expect(res.notepadAssets[0]).toEqual(asset.uuid);
		});

		it('should create a new object', () => {
			//Arrange
			// Act
			const res = notepad.addAsset(asset);

			// Assert
			expect(res).not.toBe(notepad);
		});
	});

	describe('modified', () => {
		it('should update lastModified', () => {
			// Arrange
			const expected = new Notepad('expected', { lastModified: new Date(32) });
			let notepad = new Notepad('test');

			// Act
			notepad = notepad.modified(new Date(32));

			// Assert
			expect(notepad.lastModified).toEqual(expected.lastModified);
		});
	});

	describe('toJson', () => {
		it('should generate a JSON object of the notepad', () => {
			// Arrange
			const title = 'test';

			// Act
			const n = new Notepad(title, options);

			// Assert
			expect(n.toJson()).toMatchSnapshot();
		});
	});

	describe('toXml', () => {
		it(`should generate the notepad's NPX file`, async () => {
			// Arrange
			let n = new Notepad('test', options);
			n = n.clone({
				sections: [
					n.sections[0].addNote(TestUtils.makeNote('test note'))
				]
			});

			// Act
			const res = await n.toXml();

			// Assert
			expect(res).toMatchSnapshot();
		});
	});
});

function getOptions(): NotepadOptions {
	const testSection = new Section('test section');
	(<any> testSection).internalRef = 'abc';

	return {
		lastModified: new Date(1),
		sections: [testSection],

		assets: [TestUtils.makeAsset()],
		notepadAssets: ['test']
	};
}
