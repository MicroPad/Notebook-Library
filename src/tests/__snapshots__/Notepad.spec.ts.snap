// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notepad constructor should construct with assets 1`] = `
Array [
  Asset {
    "data": Blob {
      Symbol(impl): BlobImpl {
        "_buffer": Object {
          "data": Array [
            116,
            101,
            115,
            116,
          ],
          "type": "Buffer",
        },
        "type": "",
        Symbol(wrapper): [Circular],
      },
    },
    "uuid": "abc",
  },
]
`;

exports[`Notepad constructor should construct with lastModified 1`] = `"1970-01-01T12:00:00.001+12:00"`;

exports[`Notepad constructor should construct with notepadAssets 1`] = `
Array [
  "test",
]
`;

exports[`Notepad constructor should construct with sections 1`] = `
Array [
  Section {
    "internalRef": "abc",
    "notes": Array [],
    "sections": Array [],
    "title": "test section",
  },
]
`;

exports[`Notepad flatten should generate a FlatNotepad correctly 1`] = `
FlatNotepad {
  "lastModified": "1970-01-01T12:00:00.001+12:00",
  "notepadAssets": Array [
    "test",
  ],
  "notes": Object {
    "unique": Note {
      "bibliography": Array [],
      "elements": Array [],
      "internalRef": "unique",
      "parent": "abc",
      "time": 1,
      "title": "test note",
    },
  },
  "sections": Object {
    "abc": Object {
      "internalRef": "abc",
      "title": "test section",
    },
    "deep": Object {
      "internalRef": "deep",
      "parentRef": "abc",
      "title": "nested",
    },
  },
  "title": "test",
}
`;

exports[`Notepad toJson should generate a JSON object of the notepad 1`] = `"{\\"title\\":\\"test\\",\\"lastModified\\":\\"1970-01-01T12:00:00.001+12:00\\",\\"sections\\":[{\\"title\\":\\"test section\\",\\"internalRef\\":\\"abc\\",\\"sections\\":[],\\"notes\\":[]}],\\"notepadAssets\\":[\\"test\\"]}"`;

exports[`Notepad toXml should generate the notepad's NPX file 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"no\\"?><notepad xsi:noNamespaceSchemaLocation=\\"https://getmicropad.com/schema.xsd\\" title=\\"test\\" lastModified=\\"1970-01-01T12:00:00.001+12:00\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\"><assets><asset uuid=\\"abc\\">data:;base64,dGVzdA==</asset></assets><section title=\\"test section\\"><note title=\\"test note\\" time=\\"1970-01-01T12:00:00.001+12:00\\"><addons/><bibliography/></note></section></notepad>"`;
